### PR TITLE
Add some utility functions

### PR DESCRIPTION
**Pull request by thoughtpolice**
TODO for `oscillate`:

 - [x] Should probably have a test(?)
 - [x] The type may need to be more specific, or generalized; for example it could generally take any `Bits a` type and use `complement`

I think `riseEvery` only needs a test, although it is practically very simple. I initially generalized it to take the starting value (so technically it could both function as `riseEvery` and `fallEvery`), but I figured this isn't necessary since `fmap not` will set you straight quick enough.

### CODE REVIEW

**Comment by thoughtpolice** [comment]
@christiaanb I believe this is ready to go. I refined the types to `Unsigned` to ensure that they are always synthesizable; this does have the minor negative effect you will have to constrain raw literals (e.g. you must say `riseEvery (1 :: Unsigned 1)` rather than just `riseEvery 1` because the `n` in `KnownNat n` is ambiguous). But I think this is probably fine and more appropriate anyway for Clash -- where the amount of space usage e.g. for a register is fairly clear in most cases.

**Comment by thoughtpolice** [comment]
I think I bungled one of the new tests... I'll fix it.

File: src/CLaSH/Prelude/Explicit/Safe.hs [hunk 1]
```diff
@@ -114,3 +118,23 @@ isFalling' clk is s = liftA2 edgeDetect prev s
...
+-- in order to delay a register by a known amount.
+riseEvery' :: KnownNat n => SClock clk -> Unsigned n -> Signal' clk Bool
+riseEvery' clk cnt = moore' clk transfer output 1 (pure ())
+  where
+    output   s   = if (s == cnt) then True else False
+    transfer s _ = if (s == cnt) then 1    else s+1
+
+{-# INLINEABLE oscillate' #-}
+-- | Oscillate a @'Bool'@ for a given number of cycles, given the starting state.
+oscillate' :: KnownNat n => SClock clk -> Bool -> Unsigned n -> Signal' clk Bool
```
Comment (christiaanb) [inline src/CLaSH/Prelude/Explicit/Safe.hs]: @thoughtpolice I think I prefer `SNat n` over `Unsigned n`, as you can never change the count dynamically (`Signal` doesn't have a `Monad` instance). Internally you would then use `Index n` for the counter. You would then get:

```haskell
oscillate' :: forall clk n . SClock clk -> Bool -> SNat n -> Signal' clk Bool
oscillate' clk begin cnt@SNat = moore' clk transfer snd (0, begin) (pure ())
  where
    transfer :: (Index n, Bool) -> () -> (Index n, Bool)
    transfer (s, i) _ =
      if s < snatToNum cnt
        then (s + 1,i)  -- hold current output
        else (1, not i) -- reset state and oscillate output
```

I don't think we should care too much about the probability that `< cnt` might use more logic than `== cnt`

A similar change would then have to be applied to `riseEvery'`

Comment (christiaanb) [inline src/CLaSH/Prelude/Explicit/Safe.hs]: I've changed the code in my earlier comment slightly.

**Comment by thoughtpolice** [comment]
@christiaanb Should be fixed. I think this is ready to go now.

File: src/CLaSH/Prelude/Explicit/Safe.hs [hunk 2]
```diff
@@ -114,3 +122,27 @@ isFalling' clk is s = liftA2 edgeDetect prev s
...
+
+{-# INLINEABLE riseEvery' #-}
+-- | Give a pulse every @n@ clock cycles. This is a useful helper function when
+-- combined with functions like @'CLaSH.Signal.regEn'@ or @'CLaSH.Signal.mux'@,
+-- in order to delay a register by a known amount.
+riseEvery' :: forall clk n. KnownNat n => SClock clk -> SNat n -> Signal' clk Bool
+riseEvery' clk SNat = moore' clk transfer output 0 (pure ())
+  where
+    output :: Index n -> Bool
+    output s = if (s == maxBound) then True else False
```
Comment (christiaanb) [inline src/CLaSH/Prelude/Explicit/Safe.hs]: Is the reason that this is not

```haskell
output = (== maxBound)
```

To have symmetry with `transfer` ?

Comment (thoughtpolice) [inline src/CLaSH/Prelude/Explicit/Safe.hs]: No, but I can fix it, just an oversight last night while groggy.

**Comment by thoughtpolice** [comment]
OK, finally ready.

**Comment by christiaanb** [comment]
Thank!

**Pull request merged**